airflow/models/baseoperator.py

```diff
@@ -60,9 +60,25 @@
 ScheduleInterval = Union[str, timedelta, relativedelta]
 
 
+class BaseOperatorMeta(type):
+    """
+    Base metaclass of BaseOperator.
+    """
+    def __call__(cls, *args, **kwargs):
+        """
+        Called when you call BaseOperator(). In this way we are able to perform an action
+        after initializing an operator no matter where  the ``super().__init__`` is called
+        (before or after assign of new attributes in a custom operator).
+        """
+        obj = type.__call__(cls, *args, **kwargs)
+        # Set upstream task defined by XComArgs passed to template fields of an operator
+        obj._resolve_dependants()  # pylint: disable=protected-access
+        return obj
+
+
 # pylint: disable=too-many-instance-attributes,too-many-public-methods
 @functools.total_ordering
-class BaseOperator(Operator, LoggingMixin):
+class BaseOperator(Operator, LoggingMixin, metaclass=BaseOperatorMeta):
```

Comment:
Why metaclass instead of `__init__`?

Reply:
The `resolve` method has to work for any operator, no matter where someone calls `super().__init__(*args, **kwargs)`. In other words, calling resolve in `__init__` works only when custom operator is doing:
```python
class CustomOp(BaseOperator):
    template_fields = ("custom_field",)
    def __init__(self, custom_field, *args, **kwargs):
        self.custom_field = custom_field
        super().__init__(*args, **kwargs)
```
but not in case of
```python
class CustomOp(BaseOperator):
    template_fields = ("custom_field",)
    def __init__(self, custom_field, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.custom_field = custom_field
```

This is also checked in tests.